src/com/palmergames/bukkit/towny/object/Town.java

```diff
@@ -605,31 +607,44 @@ public void setHomeBlock(@Nullable TownBlock homeBlock) {
 			setWorld(homeBlock.getWorld());
 
 		// Unset the spawn if it is not inside of the new homeblock.
-		if (spawn != null && !homeBlock.getWorldCoord().equals(WorldCoord.parseCoord(spawn))) {
+		if (spawn != null && !homeBlock.getWorldCoord().equals(Coord.parseCoord(spawn))) {
 			TownyUniverse.getInstance().removeSpawnPoint(spawn);
 			spawn = null;
 		}
 
-		if (this.hasNation() && TownySettings.getNationRequiresProximity() > 0) {
-			Nation nation = TownyAPI.getInstance().getTownNationOrNull(this);
-			if (!nation.getCapital().equals(this) && nation.getCapital().hasHomeBlock() && hasHomeBlock()) {
-				WorldCoord capitalCoord = null;
-				WorldCoord townCoord = null;
-				try {
-					capitalCoord = nation.getCapital().getHomeBlock().getWorldCoord();
-					townCoord = this.getHomeBlock().getWorldCoord();
-				} catch (TownyException ignored) {}
-				
-				if (!nation.getCapital().getHomeblockWorld().equals(getHomeblockWorld())) {
-					TownyMessaging.sendNationMessagePrefixed(nation, Translatable.of("msg_nation_town_moved_their_homeblock_too_far", getName()));
-					removeNation();
-				}
-
-				double distance = MathUtil.distance(capitalCoord.getX(), townCoord.getX(), capitalCoord.getZ(), townCoord.getZ());
-				if (distance > TownySettings.getNationRequiresProximity()) {
-					TownyMessaging.sendNationMessagePrefixed(nation, Translatable.of("msg_nation_town_moved_their_homeblock_too_far", getName()));
-					removeNation();
-				}	
+		Nation townNation = TownyAPI.getInstance().getTownNationOrNull(this);
+		if (this.hasNation() && townNation != null && !townNation.getCapital().equals(this) 
+			&& TownySettings.getNationRequiresProximity() > 0
+			&& townNation.getCapital().hasHomeBlock() && hasHomeBlock()) {
+			
+			WorldCoord capitalCoord = null;
+			WorldCoord townCoord = null;
+			try {
+				capitalCoord = townNation.getCapital().getHomeBlock().getWorldCoord();
+				townCoord = this.getHomeBlock().getWorldCoord();
+			} catch (TownyException ignored) {}
+			
+			if (!townNation.getCapital().getHomeblockWorld().equals(getHomeblockWorld())) {
+				TownyMessaging.sendNationMessagePrefixed(townNation, Translatable.of("msg_nation_town_moved_their_homeblock_too_far", getName()));
+				removeNation();
+			}
+			
+			double distance;
+			if (capitalCoord != null && townCoord != null) {
+				int x1 = capitalCoord.getX();
+				int x2 = townCoord.getX();
+				int y1 = capitalCoord.getZ();
+				int y2 = townCoord.getZ();
+				distance = MathUtil.distance(x1, x2, y1, y2);
+			} else {
+				distance = 0;
+				TownyMessaging.sendErrorMsg(String.format("(%s): Missing HomeBlock for either the Capitol or here!",
+					this.getName()));
+			}
```

Comment:
Because of the above hasHomeBlock() tests this should never cause any exceptions.

```suggestion
			WorldCoord capitalCoord = townNation.getCapital().getHomeBlockOrNull().getWorldCoord();
			WorldCoord townCoord = this.getHomeBlockOrNull().getWorldCoord();
			
			if (!townNation.getCapital().getHomeblockWorld().equals(getHomeblockWorld())) {
				TownyMessaging.sendNationMessagePrefixed(townNation, Translatable.of("msg_nation_town_moved_their_homeblock_too_far", getName()));
				removeNation();
			}
			
			int x1 = capitalCoord.getX();
			int x2 = townCoord.getX();
			int y1 = capitalCoord.getZ();
			int y2 = townCoord.getZ();
			double  distance = MathUtil.distance(x1, x2, y1, y2);
```